### PR TITLE
Check return parameter has identifier and type

### PR DESCRIPTION
**Pull request by ntrel**
Note: type cannot be `_`.
Fixes most of https://github.com/hsutter/cppfront/issues/325#issuecomment-1509817899.

### CODE REVIEW

**Review by filipsajdak — CHANGES_REQUESTED** [review]
.

**Review by filipsajdak — CHANGES_REQUESTED** [review]
.

File: source/parse.h [hunk 1]
```diff
@@ -4648,7 +4648,17 @@ class parser
             error("Cpp2 is currently exploring the path of not allowing default arguments - use overloading instead", false);
             return {};
         }
-
+        if (is_named && is_returns) {
+            auto tok = n->name();
```
Comment (hsutter) [inline source/parse.h:4652]: Might still want to `assert(tok);` too.

File: source/parse.h [hunk 2]
```diff
@@ -4648,7 +4648,17 @@ class parser
...
+            if (tok->type() != lexeme::Identifier) {
+                error("expected identifier, not '" + tok->to_string(true) + "'",
+                    false, tok->position());
+            }
+            else if (n->declaration->type.index() != declaration_node::active::a_type) {
```
Comment (hsutter) [inline source/parse.h:4657]: This can be `n->declaration->is_type()` now

**Review by hsutter — CHANGES_REQUESTED** [review]

File: regression-tests/pure2-return-tuple-no-identifier.cpp2 [hunk 3]
```diff
@@ -0,0 +1,3 @@
+d: () -> (int, int) = (1,2);
```
Comment (hsutter) [inline regression-tests/pure2-return-tuple-no-identifier.cpp2:1]: This is intended to emit an error, and test filenames that should produce errors should have names ending with `-error.cpp2`.

**Comment by ntrel** [comment]
Thanks, all 3 review points are done now.

File: source/parse.h [hunk 4]
```diff
@@ -4648,7 +4648,18 @@ class parser
...
         }
-
+        if (is_named && is_returns) {
+            auto tok = n->name();
+            assert(tok);
+            if (tok->type() != lexeme::Identifier) {
+                error("expected identifier, not '" + tok->to_string(true) + "'",
+                    false, tok->position());
+            }
+            else if (!n->declaration->is_type()) {
```
Comment (hsutter) [inline source/parse.h:4658]: Sorry, I gave an incomplete feedback on this line the first time: I think the right query is not "is this declaration a type" (a parameter shouldn't ever be a type, it should be an object) but rather "does this declaration have a concrete type"... so I think what you want is `else if (n->declaration->has_wildcard_type())`.

Did you run regressions? I think the current condition above would reject existing valid named return cases?

Comment (ntrel) [reply source/parse.h]: Sorry, last night I checked the error tests but I forgot to check the correct named return test. Now fixed.

**Comment by hsutter** [comment]
Looks good, thanks!

**Pull request merged**